### PR TITLE
feat: Add new methods: `currentIndex`, `nextIndex`, `previousIndex`, `lastIndex`, `sort`, `reverse`, and `shuffle`.

### PR DESCRIPTION
**Pull request by misteroneill**
Add new methods: `currentIndex`, `nextIndex`, `previousIndex`, `lastIndex`, `sort`, `reverse`, and `shuffle`.

## Requirements Checklist
- [x] Feature implemented / Bug fixed
- [x] If necessary, more likely in a feature request than a bug fix
  - [x] Unit Tests updated or fixed
  - [x] Docs/guides updated
- [ ] Reviewed by Two Core Contributors

### CODE REVIEW

File: src/auto-advance.js [hunk 1]
```diff
@@ -50,11 +52,14 @@ const setup = (player, delay) => {
   // Before queuing up new auto-advance behavior, check if `seconds` was
   // called with a valid value.
   if (!validSeconds(delay)) {
+    player.playlist.autoadvance_.delay = null;
     return;
   }
 
+  player.playlist.autoadvance_.delay = delay;
```
Comment (misteroneill) [inline src/auto-advance.js]: The main change in this file is to set the delay in the `setup` function. This allows us to provide more normalized values and keep ownership of the `autoadvance_` object in this module, which remaining backward-compatible.

File: src/playlist-maker.js [hunk 2]
```diff
@@ -112,7 +101,7 @@ const indexInSources = (arr, src) => {
  * @return {Function}
  *         Returns the playlist function specific to the given player.
  */
-const factory = (player, initialList, initialIndex = 0) => {
+export default function factory(player, initialList, initialIndex = 0) {
```
Comment (misteroneill) [inline src/playlist-maker.js]: Rollup 0.50 was complaining about this because `this` was undefined - it tries to more accurately mimic how native ES6 modules will work.

File: src/playlist-maker.js [hunk 3]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+  const playlist = player.playlist = (newList, newIndex = 0) => {
     if (Array.isArray(newList)) {
       list = newList.slice();
       if (newIndex !== -1) {
         playlist.currentItem(newIndex);
       }
-      playlist.changeTimeout_ = window.setTimeout(() => {
-        player.trigger('playlistchange');
-      }, 0);
+      player.setTimeout(() => player.trigger('playlistchange'), 1);
```
Comment (misteroneill) [inline src/playlist-maker.js]: Switched `setTimeout`s to player-level, so we don't need to worry about cleaning them up.

File: src/playlist-maker.js [hunk 4]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
   player.on('loadstart', () => {
     if (playlist.currentItem() === -1) {
       autoadvance.reset(player);
     }
   });
 
-  player.on('dispose', () => {
-    window.clearTimeout(playlist.changeTimeout_);
-  });
+  playlist.currentIndex_ = -1;
```
Comment (misteroneill) [inline src/playlist-maker.js]: Switching from the old `assign` to this assignment-based approach removes a largely unnecessary function and results in a slightly smaller minified file size.

File: src/playlist-maker.js [hunk 5]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
-     *
-     * @return {boolean}
-     *         The current value of repeat
+     * @event playlistsorted
+     * @type {Object}
      */
-    repeat(val) {
-      if (val !== undefined) {
-        if (typeof val !== 'boolean') {
-          videojs.log.error('Invalid value for repeat', val);
```
Comment (misteroneill) [inline src/playlist-maker.js]: Removed this strict type check as it seemed like overkill.

File: test/playlist-maker.test.js [hunk 6]
```diff
@@ -37,7 +36,7 @@ const videoList = [{
   poster: 'http://media.w3.org/2010/05/video/poster.png'
 }];
 
-QUnit.module('playlist', {
+QUnit.module('playlist-maker', {
```
Comment (misteroneill) [inline test/playlist-maker.test.js]: Name change.

Also, a few tests in this file didn't match the style of the majority of tests; so, I made them all uniform.

File: src/playlist-maker.js [hunk 7]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
 
-      for (let i = 0; i < sources.length; i++) {
-        const source = sources[i];
+  /**
+   * Get the index of the last item in the playlist.
+   *
+   * @return {number}
+   *         The index of the last item in the playlist.
+   */
+  playlist.indexOfLast = () => {
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: maybe it would be better to call this `lastIndex`?

File: src/playlist-maker.js [hunk 8]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
-          return indexInSources(list, source.src);
-        }
-      }
+  /**
+   * Get the index of the next item in the playlist.
+   *
+   * @return {number}
+   *         The index of the next item in the playlist.
+   */
+  playlist.indexOfNext = () => {
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: maybe it would be better to call this `nextIndex`?

File: src/playlist-maker.js (hunk 7 above)
Comment (BrandonOCasey) [inline src/playlist-maker.js]: maybe it would be better to call this `lastIndex`?

File: src/playlist-maker.js [hunk 9]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
 
-      playlist.currentIndex_ = -1;
-    },
+  /**
+   * Get the index of the previous item in the playlist.
+   *
+   * @return {number}
+   *         The index of the previous item in the playlist.
+   */
+  playlist.indexOfPrevious = () => {
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: `previousIndex`?

File: src/playlist-maker.js (hunk 7 above)
Comment (BrandonOCasey) [reply src/playlist-maker.js]: Should their be a `currentIndex` as well? seems to be missing. EDIT I see that currentItem returns an index rather than an item, I guess that could work, but it does seem a bit weird.

File: src/playlist-maker.js (hunk 3 above)
Comment (BrandonOCasey) [inline src/playlist-maker.js]: why was this changed to a 1? should we have a comment for why this is necessary?

File: src/playlist-maker.js [hunk 10]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+   *
+   * @param  {number} [index]
+   *         If given as a valid value, plays the playlist item at that index.
+   *
+   * @return {number}
+   *         The current item index.
+   */
+  playlist.currentItem = (index) => {
+    if (
+      typeof index === 'number' &&
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: could we just check if `index` is a `number` and if `list[index]` exists}

File: src/playlist-maker.js [hunk 11]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+      playlist.currentIndex_ !== index &&
+      index >= 0 &&
+      index < list.length
+    ) {
+      playlist.currentIndex_ = index;
+      playItem(
+        playlist.player_,
+        playlist.autoadvance_.delay,
+        list[playlist.currentIndex_]
+      );
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: Should this return, so that it works more like vjs getters/setters?

File: src/playlist-maker.js (hunk 10 above)
Comment (BrandonOCasey) [reply src/playlist-maker.js]: Would it also make sense to pass in an item from the list here and not just an index?

File: src/playlist-maker.js [hunk 12]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
-          return indexInSources(list, source);
-        } else if (source.src) {
-          return indexInSources(list, source.src);
-        }
-      }
+  /**
+   * Get the index of the next item in the playlist.
+   *
+   * @return {number}
+   *         The index of the next item in the playlist.
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: could be good to specify that it can also return `-1` if there is no next item.

File: src/playlist-maker.js [hunk 13]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+    return -1;
+  };
 
-      for (let i = 0; i < sources.length; i++) {
-        const source = sources[i];
+  /**
+   * Get the index of the last item in the playlist.
+   *
+   * @return {number}
+   *         The index of the last item in the playlist.
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: could be good to specify that it can also return `-1` if there is no next item.

File: src/playlist-maker.js [hunk 14]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
-        playlist.currentIndex_ = index;
-        playItem(
-          playlist.player_,
-          playlist.autoadvance_.delay,
-          list[playlist.currentIndex_]
-        );
-      } else {
-        playlist.currentIndex_ = playlist.indexOf(playlist.player_.currentSrc() || '');
-      }
+    return playlist.currentIndex_;
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: will this ever return `-1` like the other functions?

File: src/playlist-maker.js [hunk 15]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
-      } else {
-        nextIndex = Math.min(playlist.currentIndex_ + 1, list.length - 1);
-      }
+  /**
+   * Plays the first item in the playlist.
+   *
+   * @return {Object|undefined}
+   *         Returns undefined and has no side effects if the list is empty.
+   */
+  playlist.first = () => {
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: This name threw me off at first, I expected this function to only be a getter, perhaps it should be `playFirst`?

File: src/playlist-maker.js [hunk 16]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+   *         The number of seconds to wait before each auto-advance.
+   *         Leave this argument off to get the current auto-advance delay.
+   *
+   * @return {number|null}
+   *         The current delay.
+   */
+  playlist.autoadvance = (delay) => {
+    if (delay !== undefined) {
+      autoadvance.setup(playlist.player_, delay);
+    }
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: this does not appear to return the delay that is currently set

File: src/playlist-maker.js [hunk 17]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+   *
+   * @param  {boolean} [val]
+   *         The value to set repeat to
+   *
+   * @return {boolean}
+   *         The current value of repeat
+   */
+  playlist.repeat = (val) => {
+    if (val !== undefined) {
+      playlist.repeat_ = !!val;
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: Should we return here so that getter/setter functionality isn't mixed? I think that's what we do for video.js

File: src/playlist-maker.js (hunk 15 above)
Comment (BrandonOCasey) [reply src/playlist-maker.js]: I would think the same thing about the other similar functions

File: src/playlist-maker.js [hunk 18]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+     * Triggered after the playlist is sorted internally.
      *
-     * @return {Object|undefined}
-     *         Returns undefined and has no side effects if on first item.
+     * @event playlistsorted
+     * @type {Object}
      */
-    previous() {
+    player.trigger('playlistsorted');
+    return list;
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: It almost seems weird to just return the list here as if you want to method chain you would probably want to do so with playlist methods.

File: src/playlist-maker.js [hunk 19]
```diff
@@ -133,217 +122,336 @@ const factory = (player, initialList, initialIndex = 0) => {
...
-     *        The number of seconds to wait before each auto-advance.
+     * @event playlistsorted
+     * @type {Object}
      */
-    autoadvance(delay) {
-      playlist.autoadvance_.delay = delay;
-      autoadvance.setup(playlist.player_, delay);
-    },
+    player.trigger('playlistsorted');
+    return list;
```
Comment (BrandonOCasey) [inline src/playlist-maker.js]: same comment about the return value as above.

File: src/playlist-maker.js (hunk 3 above)
Comment (misteroneill) [reply src/playlist-maker.js]: Think I probably removed the `}, 0);` line and didn't realize it had been `0`!

File: src/playlist-maker.js (hunk 10 above)
Comment (misteroneill) [reply src/playlist-maker.js]: I dunno. I didn't want to change pre-existing API methods if I could help it.

File: src/playlist-maker.js (hunk 11 above)
Comment (misteroneill) [reply src/playlist-maker.js]: Same comment as above, I didn't want to change pre-existing API methods if I could help it.

File: src/playlist-maker.js (hunk 14 above)
Comment (misteroneill) [reply src/playlist-maker.js]: Yes, by default, and with no source on the player.

File: src/playlist-maker.js (hunk 7 above)
Comment (misteroneill) [reply src/playlist-maker.js]: It's because `currentItem()` will look at the player source to determine if it can update the `currentIndex_`; so, it's the best way to get the current index.

Comment (misteroneill) [reply src/playlist-maker.js]: As for naming... I dunno. Seems a little bike-sheddy, but the choice here was to stay consistent with the naming scheme provided by `indexOf`.

File: src/playlist-maker.js (hunk 15 above)
Comment (misteroneill) [reply src/playlist-maker.js]: I would agree, but I'm not trying to make breaking changes here. 😄

File: src/playlist-maker.js (hunk 16 above)
Comment (misteroneill) [reply src/playlist-maker.js]: Whoops.

Comment (misteroneill) [reply src/playlist-maker.js]: I'll just revert the changes to this function.

File: src/playlist-maker.js (hunk 18 above)
Comment (misteroneill) [reply src/playlist-maker.js]: Yeah, it originally didn't return anything. I'll probably stick with that pattern.

File: docs/api.md [hunk 20]
```diff
@@ -168,6 +167,112 @@ player.playlist.contains({
 // 4
 ```
 
+#### `player.playlist.indexOfLast() -> Number`
```
Comment (gkatsev) [inline docs/api.md]: 🤔  I think I would prefer these methods to be `lastIndex`, `previousIndex`, `nextIndex`.
Thoughts?

File: src/playlist-maker.js (hunk 5 above)
Comment (gkatsev) [reply src/playlist-maker.js]: it's like that due to a compromise: https://github.com/brightcove/videojs-playlist/pull/71#discussion_r105244134

File: docs/api.md (hunk 20 above)
Comment (gkatsev) [reply docs/api.md]: oh, just saw that you talked about this with @BrandonOCasey.
I'm not sure that consistency with `indexOf` is really something to strive for. `indexOf` was used to mirror the array api.
Looking to see if there are precedents for this.

Comment (misteroneill) [reply docs/api.md]: I'm cool with changing this - I don't feel strongly about the naming.

File: src/playlist-maker.js (hunk 5 above)
Comment (misteroneill) [reply src/playlist-maker.js]: That may be, but it makes no sense to me. It's inconsistent with the rest of the API and I don't see a reason why doing `repeat('true')` or `repeat(1)` shouldn't work when that kind of strictness is usually not enforced in any of our APIs...

Comment (misteroneill) [reply src/playlist-maker.js]: I'll roll it back, though. Not looking to get into a debate about API design choices.

Comment (misteroneill) [reply src/playlist-maker.js]: Rolled back. Log message will still appear.

File: src/auto-advance.js [hunk 21]
```diff
@@ -50,11 +52,14 @@ const setup = (player, delay) => {
   // Before queuing up new auto-advance behavior, check if `seconds` was
   // called with a valid value.
   if (!validSeconds(delay)) {
+    player.playlist.autoadvance_.delay = null;
```
Comment (BrandonOCasey) [inline src/auto-advance.js]: I also wonder if we should warn something here.

Comment (BrandonOCasey) [inline src/auto-advance.js]: I also wonder if we should warn something here.

Comment (misteroneill) [reply src/auto-advance.js]: The problem with that would be that you would get a warning when turning off auto-advance with a non-numeric value (e.g. `autoadvance(null)` or `autoadvance(false)`).

**Comment by misteroneill** [comment]
@gkatsev If you want another look, feel free. I'll leave the PR open.

In the meantime, I've prereleased this as `next` on npm.

File: src/playlist-maker.js [hunk 22]
```diff
@@ -133,217 +123,343 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+  /**
+   * Checks if the playlist contains a value.
+   *
+   * @param  {string|Object|Array} value
+   *         The value to check
+   *
+   * @return {boolean}
+   *         The result
+   */
+  playlist.contains = (value) => {
```
Comment (gkatsev) [inline src/playlist-maker.js]: while `contains` is a better name, should we use `includes` to match the Array and String prototypes in ES2016?

Comment (gkatsev) [reply src/playlist-maker.js]: Oh, it's a pre-existing, method, carry on.

File: src/playlist-maker.js [hunk 23]
```diff
@@ -133,217 +123,343 @@ const factory = (player, initialList, initialIndex = 0) => {
...
+        return indexInSources(list, source.src);
       }
+    }
+
+    return -1;
+  };
 
-      const sources = Array.isArray(value) ? value : value.sources;
+  /**
+   * Get the index of the current item in the playlist.
```
Comment (gkatsev) [inline src/playlist-maker.js]: can we document that this is an alias?

**Comment by misteroneill** [comment]
I just noticed an issue, I think. It looks like if you initially populate the list then call `autoadvance`, whatever value you populate will be discarded because the `autoadvance` will get wiped out when `ready` fires.
